### PR TITLE
Add Qualified Lookup Requests

### PR DESCRIPTION
**Pull request by CodaFi**
It's gonna be really confusing if this one passes but #28732 doesn't...

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

File: lib/AST/NameLookup.cpp [hunk 1]
```diff
@@ -1706,34 +1711,39 @@ bool DeclContext::lookupQualified(ArrayRef<NominalTypeDecl *> typeDecls,
...
+
+  return decls;
 }
 
 bool DeclContext::lookupQualified(ModuleDecl *module, DeclNameRef member,
                                   NLOptions options,
                                   SmallVectorImpl<ValueDecl *> &decls) const {
-  using namespace namelookup;
+  ModuleQualifiedLookupRequest req{this, module, member, options};
+  decls = evaluateOrDefault(getASTContext().evaluator, req, {});
```
Comment (hamishknight) [inline lib/AST/NameLookup.cpp]: I'm not sure we can overwrite `decls` here, from what I can tell `directReferencesForQualifiedTypeLookup` is relying on us to perform an additive lookup.

Comment (CodaFi) [reply lib/AST/NameLookup.cpp]: Oh, that's craptastic.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Review by hamishknight — APPROVED** [review]

**Comment by CodaFi** [comment]
Yup, I'm really confused now...

**Comment by CodaFi** [comment]
[Passed Compat](https://ci.swift.org/view/Pull%20Request/job/swift-PR-source-compat-suite-debug/2066/)

⛵️

**Pull request merged**